File: backend/app/schemas/agent.py
from datetime import datetime, timezone
from enum import StrEnum
from uuid import uuid4

from pydantic import BaseModel, Field, model_validator


class ToolRunStatus(StrEnum):
    planned = "planned"
    running = "running"
    completed = "completed"
    failed = "failed"


class AgentPlanRequest(BaseModel):
    scope_cidr: str = Field(description="Authorized network scope, for example 10.10.10.0/24")
    target_mode: str = Field(default="cidr", pattern="^(cidr|ip)$")
    target_ip: str | None = None
    domain: str | None = None
    dns_server: str | None = None
    rate_profile: str = Field(default="balanced", pattern="^(slow|balanced|fast)$")

    @model_validator(mode="after")
    def validate_target(self) -> "AgentPlanRequest":
        if self.target_mode == "ip" and not self.target_ip:
            raise ValueError("target_ip is required when target_mode is ip")
        return self


class AgentCommand(BaseModel):
    phase: str
    tool: str
    command: str
    purpose: str
    expected_output: str


class AgentPlan(BaseModel):
    scope_cidr: str
    target_mode: str = "cidr"
    target: str
    commands: list[AgentCommand]
    safety_notes: list[str]


class ToolRunCreate(BaseModel):
    tool: str
    command: str
    phase: str | None = None
    status: ToolRunStatus = ToolRunStatus.planned
    raw_output: str | None = None
    exit_code: int | None = None
    error: str | None = None


class ToolExecuteRequest(BaseModel):
    command: str
    scope_cidr: str
    phase: str | None = None
    timeout_seconds: int = Field(default=900, ge=5, le=7200)
    auto_ingest: bool = True


class ToolOutput(BaseModel):
    raw_output: str


class ToolRunRead(ToolRunCreate):
    id: str = Field(default_factory=lambda: str(uuid4()))
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class HostsEntry(BaseModel):
    ip_address: str
    names: list[str]

File: frontend/src/api/agent.ts
export type AgentCommand = {
  phase: string;
  tool: string;
  command: string;
  purpose: string;
  expected_output: string;
};

export type AgentPlan = {
  scope_cidr: string;
  target_mode: "cidr" | "ip";
  target: string;
  commands: AgentCommand[];
  safety_notes: string[];
};

export type AuditPhase =
  | "all"
  | "service_scan"
  | "smb_enum"
  | "ldap_enum"
  | "kerberos_enum"
  | "credential_checks"
  | "exploitation"
  | "extraction";

export type ToolRunStatus = "planned" | "running" | "completed" | "failed";

export type ToolRun = {
  id: string;
  tool: string;
  command: string;
  phase?: string | null;
  status: ToolRunStatus;
  working_directory?: string | null;
  raw_output?: string | null;
  exit_code?: number | null;
  error?: string | null;
  created_at: string;
  updated_at: string;
};

const apiBaseUrl = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000";

export async function buildAgentPlan(
  scopeCidr: string,
  domain?: string,
  targetMode: "cidr" | "ip" = "cidr",
  targetIp?: string,
  auditPhase: AuditPhase = "service_scan",
  context: {
    username?: string;
    password?: string;
    ntHash?: string;
    share?: string;
    usersList?: string;
    kaliIp?: string;
  } = {},
): Promise<AgentPlan> {
  const response = await fetch(`${apiBaseUrl}/api/agent/plan`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      scope_cidr: scopeCidr,
      target_mode: targetMode,
      target_ip: targetIp || null,
      domain: domain || null,
      audit_phase: auditPhase,
      username: context.username || null,
      password: context.password || null,
      nt_hash: context.ntHash || null,
      share: context.share || null,
      users_list: context.usersList || null,
      kali_ip: context.kaliIp || null,
      rate_profile: "balanced",
    }),
  });
  if (!response.ok) {
    throw new Error("No se pudo generar el plan del agente");
  }
  return response.json();
}

export async function executeAgentCommand(
  command: string,
  scopeCidr: string,
  phase?: string,
  workingDirectory?: string,
): Promise<ToolRun> {
  const response = await fetch(`${apiBaseUrl}/api/agent/execute`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      command,
      scope_cidr: scopeCidr,
      phase: phase || null,
      working_directory: workingDirectory || null,
      timeout_seconds: 900,
      auto_ingest: true,
    }),
  });
  if (!response.ok) {
    const error = await response.json().catch(() => null);
    throw new Error(error?.detail ?? "No se pudo ejecutar el comando");
  }
  return response.json();
}

export async function listToolRuns(): Promise<ToolRun[]> {
  const response = await fetch(`${apiBaseUrl}/api/agent/tool-runs`);
  if (!response.ok) {
    throw new Error("No se pudieron cargar las ejecuciones");
  }
  return response.json();
}

export async function cancelToolRun(runId: string): Promise<ToolRun> {
  const response = await fetch(`${apiBaseUrl}/api/agent/tool-runs/${runId}/cancel`, {
    method: "POST",
  });
  if (!response.ok) {
    throw new Error("No se pudo cancelar la ejecucion");
  }
  return response.json();
}

export async function deleteToolRun(runId: string): Promise<void> {
  const response = await fetch(`${apiBaseUrl}/api/agent/tool-runs/${runId}`, {
    method: "DELETE",
  });
  if (!response.ok) {
    throw new Error("No se pudo eliminar la ejecucion");
  }
}
